### PR TITLE
Unaryops allow empty input and API refactor

### LINKED ISSUE #2077: [BUG] unary operations don't seem to handle empty series
**Describe the bug**
There is an issue when we perform unary operations like `abs`, `min`, `max`... on an empty series.

**Steps/Code to reproduce bug**
```python
import cudf
df = cudf.DataFrame({'x':[]}
print(df.x.abs())

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-188-257386167fa5> in <module>
----> 1 df.x.abs()

/conda/envs/rapids/lib/python3.7/site-packages/cudf/dataframe/series.py in abs(self)
   1713         Returns a new Series.
   1714         """
-> 1715         return self._unaryop('abs')
   1716 
   1717     def __abs__(self):

/conda/envs/rapids/lib/python3.7/site-packages/cudf/dataframe/series.py in _unaryop(self, fn)
    454         operand.
    455         """
--> 456         outcol = self._column.unary_operator(fn)
    457         return self._copy_construct(data=outcol)
    458 

/conda/envs/rapids/lib/python3.7/site-packages/cudf/dataframe/numerical.py in unary_operator(self, unaryop)
     79     def unary_operator(self, unaryop):
     80         return numeric_column_unaryop(self, op=unaryop,
---> 81                                       out_dtype=self.dtype)
     82 
     83     def unary_logic_op(self, unaryop):

/conda/envs/rapids/lib/python3.7/site-packages/cudf/dataframe/numerical.py in numeric_column_unaryop(operand, op, out_dtype)
    406 def numeric_column_unaryop(operand, op, out_dtype):
    407     out = columnops.column_empty_like_same_mask(operand, dtype=out_dtype)
--> 408     cpp_unaryops.apply_math_op(operand, out, op)
    409     return out.view(NumericalColumn, dtype=out_dtype)
    410 

cudf/bindings/unaryops.pyx in cudf.bindings.unaryops.apply_math_op()

RuntimeError: cuDF failure at: /conda/envs/gdf/conda-bld/libcudf_1561082418487/work/cpp/src/unary/unary_ops.cuh:92: Pointer to data in input column is null
```
**Expected behavior**
We should be returning an empty series like pandas does:
```python
import pandas as pd
df = pd.DataFrame({'x':[]})
print(df.x.abs())
Series([], Name: x, dtype: float64)
```

**Environment details (please complete the following information):**
 - Environment location: Docker
 - Method of cuDF install: conda-nightly
 

**Additional context**
Found this issue when there is a dask partition with empty dataframe in it

Potential resolutions to this issue would be:
1. To handle it on python side in `numeric_column_unaryop` function guarding with a size check. 
2. Or handle it on the cpp side when the input is empty, but we seem to have an explicit null pointer check and raising exception in this line: https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/unary/unary_ops.cuh#L91 , if that's the case should the ideal approach be to fix it on python side?

cc: @kkraus14 @beckernick @randerzander

### PR DESCRIPTION
**Pull request by devavret**
Fixes #2077 and refactored the unaryops API to C++ as per @harrism 's request https://github.com/rapidsai/cudf/issues/2077#issuecomment-504876464

### CODE REVIEW

File: python/cudf/cudf/bindings/typecast.pyx [hunk 1]
```diff
@@ -7,29 +7,44 @@
 
 from .cudf_cpp cimport *
 from .cudf_cpp import *
+from cudf.bindings.unaryops cimport *
```
Comment (galipremsagar) [inline python/cudf/cudf/bindings/typecast.pyx]:
```suggestion
from cudf.dataframe.column import Column
```

File: python/cudf/cudf/bindings/typecast.pyx [hunk 2]
```diff
@@ -7,29 +7,44 @@
...
     check_gdf_compatibility(incol)
-    check_gdf_compatibility(outcol)
     
     cdef gdf_column* c_incol = column_view_from_column(incol)
-    cdef gdf_column* c_outcol = column_view_from_column(outcol)
 
-    cdef gdf_error result
+    npdtype = kwargs.get("dtype", np.float64)
+    cdef gdf_dtype dtype = dtypes[npdtype]
+
```
Comment (galipremsagar) [inline python/cudf/cudf/bindings/typecast.pyx]:
```suggestion
cdef uintptr_t category

```

**Comment by devavret** [comment]
> In the future, all such conversion logic should happen in conversion constructors in the `wrapper` types. That likely depends on #1933

Agreed. I wanted to do that in the last refactor itself but the problem of timestamp type not encapsulating the resolution (s/ms/us/ns) made me put this off. Because `gdf_timestamp` with `ms` resolution is essentially a different type from `gdf_timestamp` with `ns` resolution. The only method i could think of back then was to have the column's extra info also feed into the type dispatcher but that seemed ugly.

**Comment by jrhemstad** [comment]
> Agreed. I wanted to do that in the last refactor itself but the problem of timestamp type not encapsulating the resolution (s/ms/us/ns) made me put this off. Because `gdf_timestamp` with `ms` resolution is essentially a different type from `gdf_timestamp` with `ns` resolution. The only method i could think of back then was to have the column's extra info also feed into the type dispatcher but that seemed ugly.

Funny, that's the same conversation we were having over here: https://github.com/rapidsai/cudf/pull/2213#issuecomment-510188299

tl;dr: Instead of `gdf_time_unit`, the different resolutions will be different types, e.g., `timestamp_ns`, `timestamp_ms`, etc.

**Comment by trxcllnt** [comment]
@devavret what're your thoughts on casting between timestamps of different resolutions? can we support this now with the current `gdf_extra_dtype_info`, then update it when the new column redesign is done? I'd love to replace this [python logic](https://github.com/rapidsai/cudf/pull/2213/files#diff-cfbfb41f0663e4931eb49ed66e4c3a34R175) with a unaryop (esp. considering it's not clamping negative timestamps correctly).

**Comment by devavret** [comment]
@trxcllnt, What you want is possible when the python plumbing is added to replace the current numba approach with this PR's cast in `astype`. @galipremsagar is already on it, as part of #2295

**Comment by kkraus14** [comment]
Python / Cython side looks good to go

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -44,6 +44,21 @@ gdf_column empty_like(gdf_column const& input);
...
+ * @brief Allocates a new column of the specified size and type.
+ * 
+ * @param dtype Datatype of the column
+ * @param size Size of the column
+ * @param allocate_mask Optional Whether or not to allocate bitmask
+ * @param info Optional Extra info for the column
+ * @param stream Optional stream in which to perform allocation
+ * @return gdf_column An allocated column of given size and type
+ */
+gdf_column allocate_column(gdf_dtype dtype, gdf_size_type size,
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/include/cudf/copying.hpp]: That overload of `allocate_like` still takes the datatype from the input column. I was hoping to also use this `allocate_column` in #2156 [here](https://github.com/rapidsai/cudf/pull/2156/files#diff-98c23212fa6fcce605601c29795bb281R75). Over there I do need to allocate a column of a dtype that is not guaranteed to appear in input.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/copying.hpp]: OK, but I think this belongs in include/legacy/column.hpp (since it's not a copy and it will go away once cudf::column is deployed), not in copying.hpp.

File: cpp/src/unary/math_ops.cu [hunk 4]
```diff
@@ -198,80 +204,127 @@ private:
...
+    operator()(gdf_column const* input, gdf_column *output) {
+        CUDF_FAIL("Unsupported datatype for operation");
     }
 };
 
+} // namespace detail
+
+gdf_column unary_operation(gdf_column const& input, unary_op op) {
+
+    gdf_column output;
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu]:
```suggestion
    gdf_column output{};
```

File: cpp/src/unary/math_ops.cu [hunk 5]
```diff
@@ -198,80 +204,127 @@ private:
...
+    {
+        // TODO: replace this with a proper column constructor once
+        // cudf::column is implemented
+        bool allocate_mask = (input.valid != nullptr);
+        output = cudf::allocate_column(GDF_BOOL8, input.size, allocate_mask);
+    }
+    else
+        output = cudf::allocate_like(input);
+
+    if (input.size == 0) return output;
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu]: This should be moved above such that we return an empty column of the correct type when the input is empty.

File: cpp/src/unary/unary_ops.cuh [hunk 6]
```diff
@@ -82,41 +83,30 @@ struct Launcher {
...
+        else { // output.valid != nullptr
+            CUDA_TRY( cudaMemset(output.valid, 0xff,
+                                gdf_num_bitmask_elements( input.size )) );
+            output.null_count = 0;
         }
     }
-    else { // input->valid != nullptr
-        CUDF_EXPECTS((output->valid != nullptr),
+    else { // input.valid != nullptr
+        CUDF_EXPECTS((output.valid != nullptr),
```
Comment (jrhemstad) [inline cpp/src/unary/unary_ops.cuh]:
```suggestion
        CUDF_EXPECTS(is_nullable(output),
```

File: cpp/include/cudf/unary.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  ARCTAN,       ///< Trigonometric tangent inverse
+  EXP,          ///< Exponential (base e, Euler number)
+  LOG,          ///< Natural Logarithm (base e)
+  SQRT,         ///< Square-root (x^0.5)
+  CEIL,         ///< Smallest integer value not less than arg
+  FLOOR,        ///< largest integer value not greater than arg
+  ABS,          ///< Absolute value
+  BIT_INVERT,   ///< Bitwise Not (~)
+  NOT,          ///< Logical Not (!)
+  INVALID_UNARY ///< invalid operation
```
Comment (jrhemstad) [inline cpp/include/cudf/unary.hpp]: What's the purpose of having an `INVALID_UNARY` option? Is this needed?

Comment (devavret) [reply cpp/include/cudf/unary.hpp]: Blazing added this for whatever reason

File: cpp/src/unary/math_ops.cu (hunk 5 above)
Comment (devavret) [reply cpp/src/unary/math_ops.cu]: Actually `cudf::allocate_like` and `cudf::allocate_column` will not allocate anything if `input.size == 0`. But by keeping this line here, we can ensure that the result of `NOT` is an empty column of type `GDF_BOOL8`.

Comment (jrhemstad) [reply cpp/src/unary/math_ops.cu]: Yeah, that's a good point.
